### PR TITLE
[RW-7301][risk=low]: Logoff user before they login with eRA, also swap the order of RAS and eRA back

### PR DESCRIPTION
**Pull request by yonghaoy**
Description:
I need to deploy this to stagiung to test it because Broad Shibboleth does not integrate with staging RAS/NIH. They mocked it. 
Also I need to replace `<>` with `{}` to avoid css attacker. 

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/pages/homepage/data-access-requirements.tsx [hunk 1]
```diff
@@ -586,6 +590,7 @@ export const DataAccessRequirements = fp.flow(withProfileErrorModal)((spinnerPro
         <div style={styles.pleaseComplete}>
           Please complete the necessary steps to gain access to the <AoU/> datasets.
         </div>
+        <Iframe frameBorder={10} url={rasLogoutUrl}/>
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.tsx:593]: Should this always be rendered, or only when the user is logged in to RAS?

File: ui/src/app/utils/access-utils.tsx [hunk 2]
```diff
@@ -106,21 +106,6 @@ export const getRegistrationTasks = (navigate): RegistrationTask[] => serverConf
       return profile.twoFactorAuthCompletionTime || profile.twoFactorAuthBypassTime;
     },
     onClick: redirectToTwoFactorSetup
-  }, {
```
Comment (jmthibault79) [inline ui/src/app/utils/access-utils.tsx:109]: the order of this structure shouldn't matter - please revert

File: ui/src/app/pages/homepage/data-access-requirements.tsx [hunk 3]
```diff
@@ -540,6 +540,9 @@ const DuccCard = (props: {activeModule: AccessModule, spinnerProps: WithSpinnerO
 export const DataAccessRequirements = fp.flow(withProfileErrorModal)((spinnerProps: WithSpinnerOverlayProps) => {
   const {profile, reload} = useStore(profileStore);
 
+  // whether user finishes RAS login.
+  const [rasLoggedIn, setRasLoggedIn] = useState(false);
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.tsx:544]: this is not used

**Review by jmthibault79 — COMMENTED** [review]
Some questions:
* which option # in the doc are you implementing?
* how is the Iframe supposed to work?
* can the DAR ordering change be done in a separate PR?

File: ui/src/app/pages/homepage/data-access-requirements.tsx (hunk 1 above)
Comment (yonghaoy) [reply ui/src/app/pages/homepage/data-access-requirements.tsx:593]: ah, that is added for my testing purpose, reverted

File: ui/src/app/utils/access-utils.tsx (hunk 2 above)
Comment (yonghaoy) [reply ui/src/app/utils/access-utils.tsx:109]: I think that would affect the order in the old homepage when DAR FF is false

**Comment by yonghaoy** [comment]
> Some questions:
> 
> * which option # in the doc are you implementing?
> * how is the Iframe supposed to work?
> * can the DAR ordering change be done in a separate PR?

1: That is an alternative of option 3, just discussed in our slack thread.
2: Iframe is reverted, no longer needed. 
3: I want to do them in the same change because it is not a big change, and easy for me to push things to staging to test.

File: ui/src/app/utils/access-utils.tsx (hunk 2 above)
Comment (jmthibault79) [reply ui/src/app/utils/access-utils.tsx:109]: ah ok - that code has been removed :)

Comment (yonghaoy) [reply ui/src/app/utils/access-utils.tsx:109]: I swapped the order in my previous PR:
https://github.com/all-of-us/workbench/commit/d746ea4f39854f8b73fa73c6af38e2ad33d15b98#diff-9a246cecc6677240bbb0b36a1598e2cf3871071322f4e5b80b8b247874cfad50 so I revert it back here.
You suggest me to not revert and keep it as it?

Comment (jmthibault79) [reply ui/src/app/utils/access-utils.tsx:109]: doesn't matter.  I'm removing this soon too.

File: ui/src/app/pages/homepage/data-access-requirements.tsx [hunk 4]
```diff
@@ -35,6 +35,7 @@ import {useNavigation} from 'app/utils/navigation';
 import {profileStore, serverConfigStore, useStore} from 'app/utils/stores';
 import {AccessModule, AccessModuleStatus, Profile} from 'generated/fetch';
 import {TwoFactorAuthModal} from './two-factor-auth-modal';
+import Iframe from 'react-iframe';
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.tsx:38]: rm

File: ui/src/app/utils/access-utils.tsx [hunk 5]
```diff
@@ -60,8 +60,8 @@ export const redirectToNiH = (): void => {
   AnalyticsTracker.Registration.ERACommons();
   const url = serverConfigStore.get().config.shibbolethUiBaseUrl + '/login?return-url=' +
       encodeURIComponent(
-        window.location.origin.toString() + NIH_CALLBACK_PATH + '?token=<token>');
-  window.open(url, '_blank');
+        window.location.origin.toString() + NIH_CALLBACK_PATH + '?token={token}');
+  window.open(serverConfigStore.get().config.rasLogoutUrl + url, '_blank');
```
Comment (jmthibault79) [inline ui/src/app/utils/access-utils.tsx:64]: nit: prefer to construct the `url` including `rasLogoutUrl` so this line can be unchanged

File: ui/src/app/pages/homepage/data-access-requirements.spec.tsx [hunk 6]
```diff
@@ -94,10 +94,10 @@ describe('DataAccessRequirements', () => {
         expect(activeModule).toEqual(enabledModules[1]);
 
         // update this if the order changes
-        expect(activeModule).toEqual(AccessModule.ERACOMMONS)
+        expect(activeModule).toEqual(AccessModule.RASLINKLOGINGOV)
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.spec.tsx:97]: update line 81 to match

File: ui/src/app/pages/homepage/data-access-requirements.spec.tsx [hunk 7]
```diff
@@ -113,14 +113,11 @@ describe('DataAccessRequirements', () => {
...
     });
 
-    // TODO(RW-7301)
-    // restore the intended module order [2FA, RAS, ERA] from the temporary version [2FA, ERA, RAS]
-
     it('should return the second enabled module (RAS, not ERA) from getActiveModule' +
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.spec.tsx:119]: change to (ERA, not RAS) and make RAS the disabled module.  Update other comments here from ERA to RAS

File: ui/src/app/utils/access-utils.tsx [hunk 8]
```diff
@@ -60,8 +60,8 @@ export const redirectToNiH = (): void => {
   AnalyticsTracker.Registration.ERACommons();
   const url = serverConfigStore.get().config.shibbolethUiBaseUrl + '/login?return-url=' +
```
Comment (calbach) [inline ui/src/app/utils/access-utils.tsx:61]: FYI: you should be able to just construct this locally and enter it into your browser on staging. Also fine if you want to deploy-ui on staging though, just remember to revert.

File: ui/src/app/utils/access-utils.tsx [hunk 9]
```diff
@@ -60,8 +60,8 @@ export const redirectToNiH = (): void => {
...
       encodeURIComponent(
-        window.location.origin.toString() + NIH_CALLBACK_PATH + '?token=<token>');
-  window.open(url, '_blank');
+        window.location.origin.toString() + NIH_CALLBACK_PATH + '?token={token}');
```
Comment (calbach) [inline ui/src/app/utils/access-utils.tsx:63]: Another fix that may work or might just be necessary anyways is encoding the redirect URI again. 

i.e. 

```
window.open(serverConfigStore.get().config.rasLogoutUrl + encodeURIComponent(url), '_blank');
```

File: ui/src/app/utils/access-utils.tsx (hunk 8 above)
Comment (yonghaoy) [reply ui/src/app/utils/access-utils.tsx:61]: RAS does not add localhost:4200 into their the prod client id allowlist. I need to deploy-ui on staging using this PR after this week release

File: ui/src/app/utils/access-utils.tsx (hunk 9 above)
Comment (yonghaoy) [reply ui/src/app/utils/access-utils.tsx:63]: do you mean keep using` <>`, but encode "%3C and "%3E again?

File: ui/src/app/pages/homepage/data-access-requirements.tsx [hunk 10]
```diff
@@ -571,7 +570,7 @@ export const DataAccessRequirements = fp.flow(withProfileErrorModal)((spinnerPro
     setActiveModule(getActiveModule(enabledModules, profile));
   }, [profile]);
 
-  const {config: {unsafeAllowSelfBypass}} = useStore(serverConfigStore);
+  const {config: {unsafeAllowSelfBypass, rasLogoutUrl}} = useStore(serverConfigStore);
```
Comment (jmthibault79) [inline ui/src/app/pages/homepage/data-access-requirements.tsx:573]: unused

**Review by jmthibault79 — APPROVED** [review]

File: ui/src/app/utils/access-utils.tsx (hunk 9 above)
Comment (calbach) [reply ui/src/app/utils/access-utils.tsx:63]: If {} works, that's fine. Yes, with <> escape twice is what I meant

File: ui/src/app/utils/access-utils.tsx (hunk 8 above)
Comment (calbach) [reply ui/src/app/utils/access-utils.tsx]: moot point now, but I meant either locally or manually, generating a URL for **staging**, not localhost; then just pasting into the browser.

**Review by calbach — APPROVED** [review]

**Pull request merged**